R/MsBackendMemory.R

```diff
@@ -0,0 +1,657 @@
+          function(object, spectraVariables = spectraVariables(object)) {
+              if (!all(spectraVariables %in% spectraVariables(object)))
+                  stop("Spectra variables ",
+                       paste(spectraVariables[!(spectraVariables %in%
+                                                spectraVariables(object))],
+                             collapse = ", "), " not available")
+              ## spectraData
+              keep <- intersect(spectraVariables, colnames(object@spectraData))
+              object@spectraData <- object@spectraData[, keep, drop = FALSE]
+              ## peaksData
+              if (length(object@peaksData)) {
+                  have <- colnames(object@peaksData[[1L]])
+                  keep <- intersect(spectraVariables, have)
+                  if (length(keep) < length(have))
+                      object@peaksData <- lapply(object@peaksData, function(z)
+                          z[, keep, drop = FALSE])
+              }
+              ## peaksDataFrame
+              if (length(object@peaksDataFrame)) {
+                  have <- colnames(object@peaksDataFrame[[1L]])
+                  keep <- intersect(spectraVariables, have)
+                  if (!length(keep)) {
+                      object@peaksDataFrame <- list()
+                  } else {
+                      if (length(keep) < length(have))
+                          object@peaksDataFrame <- lapply(
+                              object@peaksDataFrame, function(z)
+                                  z[, keep, drop = FALSE])
+                  }
+              }
+              msg <- .valid_spectra_data_required_columns(object@spectraData)
+              if (length(msg))
+                  stop(msg)
+              validObject(object)
+              object
+})
+
+#' @rdname hidden_aliases
+setMethod("smoothed", "MsBackendMemory", function(object) {
+    .get_column(object@spectraData, "smoothed")
+})
+
+#' @rdname hidden_aliases
+#'
+#' @aliases smoothed<-,MsBackendMemory-method
+setReplaceMethod("smoothed", "MsBackendMemory", function(object, value) {
+    value_len <- length(value)
+    value_type <- is.logical(value)
+    if (value_type && (value_len == 1L || value_len == length(object)))
+        object@spectraData$smoothed <- value
+    else
+        stop("'value' has to be a 'logical' of length 1 or ", length(object))
+    validObject(object)
+    object
+})
+
+#' @rdname hidden_aliases
+#'
+#' @importFrom methods as
+#'
+#' @importFrom S4Vectors SimpleList
+#'
+#' @importMethodsFrom S4Vectors lapply
+setMethod(
+    "spectraData", "MsBackendMemory",
+    function(object, columns = spectraVariables(object)) {
+        as(.df_spectra_data(object, columns), "DataFrame")
+    })
+
+#' @rdname hidden_aliases
+#'
+#' @note: this replaces **all** the data in the backend.
+setReplaceMethod("spectraData", "MsBackendMemory", function(object, value) {
+    if (inherits(value, "DataFrame")) {
```

Comment:
Shouldn't we allow a `data.frame` here as well?

Looks wired:

```r
d <- data.frame(a = 1:3)
# internally it is converted in a data.frame
spectraData(msbckmem) <- as(d, "DataFrame")
# internally it is converted from a data.frame to a DataFrame
all.equal(as.data.frame(spectraData(msbckmem)), d)
```

Reply:
yes, you're right. Maybe best to support `data.frame` and `DataFrame`.

Reply:
thinking it over, no, I don't think it's a good idea. This would break the convention of the `MsBackend` interface. All backends return a `DataFrame` with `spectraData` and are supposed to take a `DataFrame` as input. I would not want to change this now for this one backend.